src/cache/cache-localstorage.ts

```diff
@@ -0,0 +1,68 @@
+import {
+  CacheKey,
+  findExistingCacheKey,
+  ICache,
+  CACHE_KEY_PREFIX
+} from './shared';
+
+export class LocalStorageCache implements ICache {
+  public set(key: string, entry: unknown): Promise<void> {
+    localStorage.setItem(key, JSON.stringify(entry));
+    return Promise.resolve();
+  }
+
+  public get(key: string): Promise<unknown> {
+    return new Promise(resolve => {
+      const cacheKey = CacheKey.fromKey(key);
+      const payload = this.readJson(cacheKey);
+
+      if (!payload) resolve(null);
+
+      resolve(payload);
+    });
+  }
+
+  public remove(key: string): Promise<void> {
+    localStorage.removeItem(key);
+    return Promise.resolve();
+  }
+
+  public clear(): Promise<void> {
+    return new Promise(resolve => {
+      for (var i = localStorage.length - 1; i >= 0; i--) {
+        if (localStorage.key(i).startsWith(CACHE_KEY_PREFIX)) {
+          localStorage.removeItem(localStorage.key(i));
+        }
+      }
+
+      resolve();
+    });
+  }
+
+  /**
+   * Retrieves data from local storage and parses it into the correct format
+   * @param cacheKey The cache key
+   */
+  private readJson(cacheKey: CacheKey): unknown {
+    const existingCacheKey = findExistingCacheKey(
+      cacheKey,
+      Object.keys(window.localStorage)
+    );
```

Comment:
@adamjmcgrath I ended up not extracting this out at this stage as I think it needs more thought. The only way we can really push this logic into the manager is to have a cache implement an `all` or `keys` function but realistically most cache implementations that would be useful for the primary use case (Ionic, iOS, etc) will likely use some kind of keychain plugin which don't appear to provide this.

Happy to dive into this further though as I don't think this `findExistingCacheKey` logic belongs here either.

Reply:
I thought we were discussing changing the data structure, so that If the cache implementation you're wrapping can't be iterated over (no `all` or `keys`), a developer doesn't have to come up with their own equivalent of `findExistingCacheKey`?


```js
// old
'prefix::client_id::audience::scopeA': 'tokenA'
'prefix::client_id::audience::scopeB': 'tokenB'

// new
'prefix::client_id::audience': { scopeA: 'tokenA', scopeB: 'tokenB' }
```

With the caveat that we keep `localStorage` the same so we don't log users out or break Cypress

Reply:
We did, didn't we 🤔  I don't think I was clear/convinced about the solution to that but now that you've illustrated it here it makes more sense. Let me revisit.

Reply:
👍 Let me know if you want to pair on it